### ISSUE TITLE
Feature Request: Stroke smoothing

### ISSUE THREAD

**Issue by ZackMercury** [serge-rgb/milton#122]
Screenshot from Milton
![image](https://user-images.githubusercontent.com/10474828/58654606-effc6b00-8320-11e9-9b79-b94bdec59b01.png)
(jagged, ugly strokes)

Screenshot from Mischief Free (limited version)
![image](https://user-images.githubusercontent.com/10474828/58654775-5b463d00-8321-11e9-93ad-ba9936e71218.png)
(strokes are perfectly smooth on all levels of zoom)

I believe Milton has big future, but only if you find a way to fix those ugly strokes.
By no means I know how to do it, but I assume Mischief does conversion to bezier curves and averages thickness on the intervals and renders them.

**Comment by serge-rgb** [comment]
Totally agree.. I made a failed attempt at better smoothing a month or two ago, but ran into problems with the approach I was taking. There's still a lot that can be done though.

**Comment by Apostolique** [comment]
As far as I can tell, my strokes look better than that. 
![image](https://user-images.githubusercontent.com/2766423/58657202-c5dc8600-82eb-11e9-821c-89f7ae400547.png)

Could use a bit of smoothing, but otherwise it looks quite good to me.

**Comment by ZackMercury** [comment]
> As far as I can tell, my strokes look better than that.
> ![image](https://user-images.githubusercontent.com/2766423/58657202-c5dc8600-82eb-11e9-821c-89f7ae400547.png)
> 
> Could use a bit of smoothing, but otherwise it looks quite good to me.

1) Wait till you zoom in.
2) You don't write text that big, usually you use little space of your screen cause they might be just a small part of your problem solution, but they still should look smooth and pretty.

**Comment by ZackMercury** [comment]
Now let me zoom in in Mischief
![image](https://user-images.githubusercontent.com/10474828/58657648-6c924800-8327-11e9-93a2-0a499ce14450.png)

**Comment by ZackMercury** [comment]
Yes, it does conversion to smoothed beziers on the fly, while you draw.
I just tried to draw a stroke really quickly and saw jagged line edges, which instantaneously were replaced by a smooth curve.

**Comment by ZackMercury** [comment]
Maybe it also takes your stroke velocity at all the given points into account to smooth them correctly.
For example, if you were too loose, meaning, drew some fast strokes with a lot of extra movements, it should smooth them more than those that are with a smaller drawing velocity.

**Comment by Apostolique** [comment]
I zoomed in on the 2:
![image](https://user-images.githubusercontent.com/2766423/58658733-27522400-82ef-11e9-89e6-a90d10c42437.png)

Or this (zoomed in quite a lot): 
![image](https://user-images.githubusercontent.com/2766423/58659110-1e158700-82f0-11e9-9678-a67c3f692909.png)

I don't see the same sort of artifacts that you see in your first post.

Anyway, last time smoothing was done in Milton, it was done through a shader iirc. It looked quite good. The only downside was that it broke straight lines. If we can get a hybrid system, then it could work.

**Comment by ZackMercury** [comment]
A hybrid system, you mean a Pen tool for smooth lines and Pencil tool for straight lines? :P

**Comment by serge-rgb** [comment]
@Apostolique Yup! I thought I could do bezier interpolation in the shader, but failed miserably :). You're right in that the problem came to light after I broke straight lines, but the reason is a bit more complicated. Basically, Milton shades pixels by projecting each pixel's center to its corresponding line segment(s) of the stroke and comparing its distance with the brush radius. 
The problem with doing Bezier in the shader is that projecting arbitrary points to a bezier curve is not a trivial problem. I am limited in (1) my math skills, (2) the time I can devote to Milton and (3) there's always perf concerns; doing tons of math in the fragment shader is a Bad Idea.

That said, there are still things I want to try out to make strokes prettier. I don't know if Milton's strokes will ever be as pretty as Mischief's, but at the very least there are some incremental improvements I can make that might get it some bit of the way there.

**Comment by Apostolique** [comment]
The line tool is already a separate option. You can press L to use the line tool.

This is how the 2 looks with smoothing enabled: 
![image](https://user-images.githubusercontent.com/2766423/58659879-10610100-82f2-11e9-94ca-1fef8d9c3784.png)

Or this: 
![image](https://user-images.githubusercontent.com/2766423/58659963-3c7c8200-82f2-11e9-9d15-6ff79fd088fa.png)

This is how it broke the lines last time: https://gfycat.com/SoggyDesertedDingo

_______

To do what you want to do, Milton would have to clean up the strokes themselves which might not be bad, but there's room to lose fidelity.

Personally, I would add a drop down menu with different smoothing algorithms that could be applied. But already, this is sounding out of scope for Milton.

**Comment by ZackMercury** [comment]
Something I could google
http://willowsystems.github.io/jSignature/%23%2Fabout%2Flinesmoothing%2F.html
https://www.tandfonline.com/doi/abs/10.1080/2151237X.2008.10129260

**Comment by ZackMercury** [comment]
Oh, there you go, I found a one with free access https://pdfs.semanticscholar.org/9f5d/fa77f61a4dc87faa243f995e2092ddb3f521.pdf

**Comment by aameen951** [comment]
@serge-rgb I noticed that strokes in `v1.6.0` and forward are not as smooth as they were before. Specifically after the commit https://github.com/serge-rgb/milton/commit/0af10f85806186dfcfc7287c8f652ccb75926789 where Milton switched from using Bezier to a different algorithm and in `v1.6.2` smoothing was completely disabled.

I would like to try and do something to smooth strokes again and it would be useful to know why it was removed.

**Comment by aameen951** [comment]
I started working on stroke smoothing and created a debug tool in Milton to allow me to capture a stroke and try different algorithms.

The first thing I did is visualize what the current algorithm is doing (which averages the last n points). The following screenshot show the raw stroke (without any modifications) and the stroke after applying the algorithm:
![y1L58jHui9](https://user-images.githubusercontent.com/7351160/65015030-11618100-d929-11e9-8a84-fa5854a79678.gif)

It looks smoother after applying the algorithm but the stroke has moved quite a bit from the raw stroke and, more importantly, the last point in the stroke is behind the raw point by a distance.

**Comment by serge-rgb** [comment]
That is pretty cool!
I wrote a bit on this thread about why I backtracked on the Bezier approach.

What I believe now is, as long as Milton uses piecewise linear strokes (i.e. not doing Bezier interpolation in the shader), the real problem is going to be stroke resolution, not stroke smoothing. I.e. the smoothing helps deal with the noise in the input signal, but it can do nothing about the fact that there just aren't enough points, so doing piecewise linear stroke segments will always show those ugly segment joints.

Even in the cool gif you just posted, if I were to zoom into the smoothed stroke, I could easily zoom into a "joint" and notice the angle between two stroke segments.

Going forward Milton will stay with piecewise linear strokes, to keep the pixel shader simple. However, there is still the option of adding interpolation points in between the stroke points. Those interpolation points can be dynamically generated on a LOD basis based on the relative size of the stroke to the screen. I.e. Larger (relative to the screen) strokes get more interpolation points. Smaller (relative to the screen) strokes get fewer or no interpolation points (and might even discard some real stroke points if we want to go that far. I don't it needs to go that far, since the rendering bottleneck probably won't ever be the vertex stage). The number of interpolation points can be chosen heuristically based on what looks good.

So, summing up. Yes, strokes are still ugly. No, it's not the biggest renderer problem. The biggest renderer problem is that Milton is driver-bound at the moment, and my next many hours of hacking on Milton will be working toward improving that. The solution to the ugly strokes is not going to be a better smoothing algorithm though, it's going to be dynamic interpolation points at cook time.

I would like to get to interpolation points eventually, but it might not be soon. Looks like you're already comfortable hacking Milton. If you want to add support for adding interpolation points at stroke-cook time, I would cheer you on and help you with any questions! If not, that's totally fine, I will get to it eventually when other higher priority items get crossed off the list.

**Comment by aameen951** [comment]
My understanding of the problem is adding more points will make the line smoother only at the current zoom and as you zoom in more it's no longer smooth.

My understanding of the solution is that points need to be fitted into a curve (and stored as a curve), and that curve (at render time) is tessellated on the CPU at the current zoom with enough points to stay smooth. This way your still submitting lines to the shader, storing the points efficiently, and the strokes are smooth regardless of the zoom.

Am I correct?

**Comment by serge-rgb** [comment]
My proposed solution is the first one, just adding more points to make it "smoother", and doing it dynamically on zoom, so zooming in would cause strokes to get more points.

The second one would also be a solution, but it would also be a rewrite of a big chunk of Milton :)

**Comment by ZackMercury** [comment]
Above I sent you a paper which as I believe is currently used by most drawing applications and produces a nice result like Mischief does. 
I'm sure it will help a lot. 
Storing strokes as beziers is definitely a step forward since then you can zoom in infinitely and add as many points as you like, you can make a setting for that as in the max amount of points that you can have in the viewport, and add points along beziers based off that.

**Comment by aameen951** [comment]
@ZackMercury thank you very much. I'll definitely take a look at that.

I don't see a download button in the provided link, but I found an alternative one:
https://web.archive.org/web/20170630045714/http://www.sarahfrisken.com/efficientCurveFitting.pdf

I'll be posting here on the progress soon.

**Comment by ZackMercury** [comment]
https://pdfs.semanticscholar.org/9f5d/fa77f61a4dc87faa243f995e2092ddb3f521.pdf

**Comment by ZackMercury** [comment]
Oh, sorry, it was a direct link to the article. For some reason they removed it.

**Comment by aameen951** [comment]
I implemented this paper by Sarah Frisken and this is the result:

![tcaR7u2UJF](https://user-images.githubusercontent.com/7351160/67545616-6e0f4300-f702-11e9-889b-57fd57327f7f.gif)

Unfortunately, it is not that good. The reason is it tries to find _a_ curve that fit the points not a better curve due to the optimization it's doing (or I did something wrong).
I'll look for other algorithms, If anyone has any suggestions or if you know of an open source software that has smooth pen drawing please drop me a comment.

**Comment by ZackMercury** [comment]
can you try and feed the algorithm with less points

**Comment by ZackMercury** [comment]
you want maxerror at something like 1.5-2 and less control points.

**Comment by aameen951** [comment]
Yes I can, but this is what a normal tablet will generate.

![hIhTkpl6nU](https://user-images.githubusercontent.com/7351160/67622381-44e5d400-f822-11e9-8e35-892ba51c588b.gif)

**Comment by ZackMercury** [comment]
hmm I really don't like that wavy piece. did you use the code from the paper?

**Comment by ZackMercury** [comment]
we can try some other curve fitting methods out there then.
since you implemented beziers in milton, it's just a matter of 30 mins to add a new algo

**Comment by aameen951** [comment]
It's not exactly like the paper, I changed how you adjust the curve when adding new points (before starting the fitting process) because the results were worse.

**Comment by ZackMercury** [comment]
I can just google for a bit and drop a couple potentially useful links here
http://paperjs.org/tutorials/paths/smoothing-simplifying-flattening/

**Comment by aameen951** [comment]
Yes, adding more algorithms is not that hard. I built debugging stuff that makes it easier to experiment with the algorithm.
Bezier curves are just a hack, it's a function that split the curve into small straight lines.

**Comment by aameen951** [comment]
I'll take a look.

I also found this book:

> Least Squares Orthogonal Distance Fitting of Curves and Surfaces in Space

I'll try to get it and see if it contains good algorithms.

**Comment by ZackMercury** [comment]
well that one might be too slow to use realtime, you know these are math problems distantly related to computer graphics, so better check this path.simplify method of the one I sent. it looks interesting.
also path.fit https://github.com/paperjs/paper.js/blob/develop/src/path/PathFitter.js

**Comment by ZackMercury** [comment]
and path.fitCubic

**Comment by aameen951** [comment]
I'll definitely check that. Thanks!

**Comment by Apostolique** [comment]
Similar discussion here: https://github.com/opentoonz/opentoonz/issues/81

**Comment by ZackMercury** [comment]
That looks sick.

**Comment by ZackMercury** [comment]
Any updates on stabilization and smoothing?
